### PR TITLE
added xkcd functionality

### PR DESCRIPTION
**Pull request by Mole1424**
added xkcd functionality as per https://github.com/UWCS/apollo/issues/117

### CODE REVIEW

File: cogs/commands/xkcd.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+SHORT_HELP_TEXT = "For all your xkcd needs"
+
+
+class XKCD(commands.Cog):
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @commands.hybrid_command(help=LONG_HELP_TEXT, brief=SHORT_HELP_TEXT)
+    async def xkcd(self, ctx: Context, comic_id: int = -1):
```
Comment (Joeyh021) [inline cogs/commands/xkcd.py:29]: The argument here should probably be `comid_id: int | None = None`. Using sentinal values is bad.

File: cogs/commands/xkcd.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        """gets either a random comic or a specific one"""
+        max_comic_id = await self.get_recent_comic()  # gets the most recent comic's id
+        if (
+            comic_id <= 0 or comic_id > max_comic_id
+        ):  # if invalid id then generate a random valid one
```
Comment (Joeyh021) [inline cogs/commands/xkcd.py:34]: imo throwing an error to the user that their comic doesn't exist is probably better than returning a random one

File: cogs/commands/xkcd.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            comic_id = random.randint(1, max_comic_id)
+
+        comic_return = await self.get_comic(comic_id)  # get the raw json of the comic
+        if comic_return == None:
```
Comment (Joeyh021) [inline cogs/commands/xkcd.py:38]: nitpick but `None` is falsy: you can do `if not comic_return`

File: cogs/commands/xkcd.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        comic_json = json.loads(await self.get_comic(comic_id))  # convert into readable
+        comic_img = await self.get_comic_image(comic_json["img"])
+        if comic_img == None:
+            return await ctx.reply("could not get comic image")
+        await ctx.reply(
+            comic_json["safe_title"], file=comic_img
+        )  # reply with comic title and image
+
+    async def get_comic(self, comic_id: int):
```
Comment (Joeyh021) [inline cogs/commands/xkcd.py:49]: return type hint needed

File: cogs/commands/xkcd.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        """gets the most recent comic id"""
+        async with aiohttp.ClientSession() as session:
+            async with session.get("https://xkcd.com/info.0.json") as response:
+                if response.status == 200:
+                    logging.info("successfully got moset recent comic")
+                    xkcd_response = json.loads(await response.read())
+                    return xkcd_response["num"]
+                else:
+                    logging.info("failed to get comic")
+                    return -1
```
Comment (Joeyh021) [inline cogs/commands/xkcd.py:72]: don't return -1, mark the return type as `int | None` and return None

File: cogs/commands/xkcd.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    async def get_comic_image(self, url):
```
Comment (Joeyh021) [inline cogs/commands/xkcd.py:74]: return type, `url` argument type

**Comment by Joeyh021** [comment]
Other than those comments, most of which is my crusade against untyped code, fantastic work!

File: cogs/commands/xkcd.py (hunk 3 above)
Comment (ericthelemur) [reply cogs/commands/xkcd.py:38]: There is no xkcd 0, but i'd prefer `is None` as it feels safer. Also technically safer to use `is` over of `==`.

**Pull request merged**